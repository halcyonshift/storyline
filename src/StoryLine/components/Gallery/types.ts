export type ImageType = {
    path: string
    title?: string
    subtitle?: string
}

export type GalleryProps = {
    images: ImageType[]
    layout?: 'horizontal' | 'vertical'
}
