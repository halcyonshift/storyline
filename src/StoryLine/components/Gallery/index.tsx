import { useEffect, useState, SyntheticEvent } from 'react'
import {
    Backdrop,
    Box,
    IconButton,
    ImageList,
    ImageListItem,
    ImageListItemBar,
    Stack
} from '@mui/material'
import { useTranslation } from 'react-i18next'
import { GLOBAL_ICONS } from '@sl/constants/icons'
import useLayout from '@sl/layouts/Work/useLayout'
import Image from '../Image'
import { GalleryProps } from './types'

const Gallery = ({ images, layout }: GalleryProps) => {
    const [index, setIndex] = useState<number>(0)
    const [open, setOpen] = useState<boolean>(false)
    const [imageHeight, setImageHeight] = useState<number>(0)
    const { panelWidth, navigationWidth, windowWidth } = useLayout()
    const { t } = useTranslation()

    const nextImage = (e: SyntheticEvent) => {
        e.stopPropagation()
        setIndex(index + 1 === images.length ? 0 : index + 1)
    }

    const lastImage = (e: SyntheticEvent) => {
        e.stopPropagation()
        setIndex(index - 1 < 0 ? images.length - 1 : index - 1)
    }

    useEffect(() => {
        if (layout === 'vertical') {
            setImageHeight(Math.round((windowWidth - panelWidth - navigationWidth) / 3 / 3))
        } else {
            setImageHeight(
                Math.round((windowWidth - panelWidth - navigationWidth) / 2 / 3) -
                    images.length * 3 -
                    3
            )
        }
    }, [panelWidth, navigationWidth, windowWidth])

    if (!images.length) return <></>

    return (
        <Box>
            <Backdrop className='z-10' open={open} onClick={() => setOpen(false)}>
                <Stack direction='row' className='w-[80%]' alignItems='center'>
                    {images.length > 1 ? (
                        <IconButton
                            aria-label={t('component.gallery.back')}
                            onClick={lastImage}
                            className='text-white'>
                            {GLOBAL_ICONS.back}
                        </IconButton>
                    ) : null}
                    <Box className='flex-grow'>
                        <Image path={images[index].path} className='mx-auto'></Image>
                    </Box>
                    {images.length > 1 ? (
                        <IconButton
                            aria-label={t('component.gallery.next')}
                            onClick={nextImage}
                            className='text-white'>
                            {GLOBAL_ICONS.next}
                        </IconButton>
                    ) : null}
                </Stack>
            </Backdrop>
            <Box className={layout === 'vertical' ? '' : 'grid gap-1 grid-cols-2'}>
                <Image path={images[index].path} onClick={() => setOpen(true)}></Image>
                <Box>
                    <ImageList cols={3} className={layout == 'vertical' ? 'my-1' : ''}>
                        {images.map((image, i) => (
                            <ImageListItem key={`image-${image.path}`} onClick={() => setIndex(i)}>
                                <Image
                                    path={`${image.path}`}
                                    alt={image.title}
                                    style={{
                                        height: imageHeight,
                                        objectFit: 'cover'
                                    }}
                                />
                                {image.title ? (
                                    <ImageListItemBar
                                        title={image.title}
                                        subtitle={image.subtitle}
                                    />
                                ) : null}
                            </ImageListItem>
                        ))}
                    </ImageList>
                </Box>
            </Box>
        </Box>
    )
}

export default Gallery
