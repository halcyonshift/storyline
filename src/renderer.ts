import { init } from '@sentry/electron/renderer'
import { init as reactInit } from '@sentry/react'

if (parseInt(process.env.DEBUG) === 0) {
    init({ dsn: process.env.SENTRY }, reactInit)
}

import '@fontsource/roboto/300.css'
import '@fontsource/roboto/400.css'
import '@fontsource/roboto/500.css'
import '@fontsource/roboto/700.css'
import './index.css'
import './StoryLine'
