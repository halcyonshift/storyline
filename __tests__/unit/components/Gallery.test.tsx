import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import Gallery from '@sl/components/Gallery'
import { render, screen } from '../test-utils'

describe('<Gallery />', () => {
    it('displays nothing if no images given', () => {
        render(<Gallery images={[]} />)
        expect(screen.queryByAltText('an image')).toBeFalsy()
    })

    it('displays title and subtitle if given', async () => {
        render(
            <Gallery
                images={[{ path: 'image', title: 'An Image', subtitle: 'A subtitle' }]}
                layout='vertical'
            />
        )

        const title = await screen.findByText('An Image')
        const subtitle = await screen.findByText('A subtitle')
        expect(title).toBeTruthy()
        expect(subtitle).toBeTruthy()
    })

    it('navigates forward', async () => {
        render(
            <Gallery
                images={[
                    { path: 'image1.jpg', title: 'Image 1' },
                    { path: 'image2.jpg', title: 'Image 2' }
                ]}
            />
        )

        const next = await screen.findByLabelText('component.gallery.next')

        const image1 = await screen.findByText('Image 1')
        expect(image1).toBeTruthy()

        await userEvent.click(next)

        const image2 = await screen.findByText('Image 2')
        expect(image2).toBeTruthy()

        await userEvent.click(next)

        const image3 = await screen.findByText('Image 1')
        expect(image3).toBeTruthy()
    })

    it('navigates back', async () => {
        render(
            <Gallery
                images={[
                    { path: 'image1.jpg', title: 'Image 1' },
                    { path: 'image2.jpg', title: 'Image 2' }
                ]}
            />
        )

        const back = await screen.findByLabelText('component.gallery.back')

        const image1 = screen.findByText('Image 1')
        expect(image1).toBeTruthy()

        await userEvent.click(back)

        const image2 = screen.findByText('Image 2')
        expect(image2).toBeTruthy()

        await userEvent.click(back)

        const image3 = screen.findByText('Image 1')
        expect(image3).toBeTruthy()
    })

    it('renders correctly', () => {
        const images = [{ path: 'image', title: 'An Image', subtitle: 'A subtitle' }]

        const verticalTree = renderer.create(<Gallery images={images} layout='vertical' />).toJSON()
        expect(verticalTree).toMatchSnapshot()

        const horizontalTree = renderer.create(<Gallery images={images} />).toJSON()
        expect(horizontalTree).toMatchSnapshot()
    })
})
