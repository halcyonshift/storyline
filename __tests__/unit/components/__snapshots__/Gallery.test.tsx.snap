// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gallery /> renders correctly 1`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    aria-hidden={true}
    className="MuiBackdrop-root z-10 css-sox5kk-MuiBackdrop-root"
    onClick={[Function]}
    style={
      {
        "opacity": 0,
        "visibility": "hidden",
      }
    }
  >
    <div
      className="w-[80%] css-1d9cypr-MuiStack-root"
    >
      <div
        className="flex-grow MuiBox-root css-0"
      />
    </div>
  </div>
  <div
    className="MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-0"
    >
      <ul
        className="MuiImageList-root MuiImageList-standard my-1 css-rvv586-MuiImageList-root"
        style={
          {
            "gap": 4,
            "gridTemplateColumns": "repeat(3, 1fr)",
          }
        }
      >
        <li
          className="MuiImageListItem-root MuiImageListItem-standard css-ta95xa-MuiImageListItem-root"
          onClick={[Function]}
          style={
            {
              "gridColumnEnd": "span 1",
              "gridRowEnd": "span 1",
              "height": "auto",
              "marginBottom": undefined,
            }
          }
        >
          <div
            className="MuiImageListItemBar-root MuiImageListItemBar-positionBottom css-ztrqiq-MuiImageListItemBar-root"
          >
            <div
              className="MuiImageListItemBar-titleWrap MuiImageListItemBar-titleWrapBottom css-1lizpae-MuiImageListItemBar-titleWrap"
            >
              <div
                className="MuiImageListItemBar-title css-dasnyc-MuiImageListItemBar-title"
              >
                An Image
              </div>
              <div
                className="MuiImageListItemBar-subtitle css-hk6gwf-MuiImageListItemBar-subtitle"
              >
                A subtitle
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<Gallery /> renders correctly 2`] = `
<div
  className="MuiBox-root css-0"
>
  <div
    aria-hidden={true}
    className="MuiBackdrop-root z-10 css-sox5kk-MuiBackdrop-root"
    onClick={[Function]}
    style={
      {
        "opacity": 0,
        "visibility": "hidden",
      }
    }
  >
    <div
      className="w-[80%] css-1d9cypr-MuiStack-root"
    >
      <div
        className="flex-grow MuiBox-root css-0"
      />
    </div>
  </div>
  <div
    className="grid gap-1 grid-cols-2 MuiBox-root css-0"
  >
    <div
      className="MuiBox-root css-0"
    >
      <ul
        className="MuiImageList-root MuiImageList-standard css-rvv586-MuiImageList-root"
        style={
          {
            "gap": 4,
            "gridTemplateColumns": "repeat(3, 1fr)",
          }
        }
      >
        <li
          className="MuiImageListItem-root MuiImageListItem-standard css-ta95xa-MuiImageListItem-root"
          onClick={[Function]}
          style={
            {
              "gridColumnEnd": "span 1",
              "gridRowEnd": "span 1",
              "height": "auto",
              "marginBottom": undefined,
            }
          }
        >
          <div
            className="MuiImageListItemBar-root MuiImageListItemBar-positionBottom css-ztrqiq-MuiImageListItemBar-root"
          >
            <div
              className="MuiImageListItemBar-titleWrap MuiImageListItemBar-titleWrapBottom css-1lizpae-MuiImageListItemBar-titleWrap"
            >
              <div
                className="MuiImageListItemBar-title css-dasnyc-MuiImageListItemBar-title"
              >
                An Image
              </div>
              <div
                className="MuiImageListItemBar-subtitle css-hk6gwf-MuiImageListItemBar-subtitle"
              >
                A subtitle
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
